upstairs/src/lib.rs

```diff
@@ -1555,44 +1536,86 @@ impl Downstairs {
+                                    &response.data[..],
+                                ]);
+
+                                if computed_hash == response.hashes[i] {
+                                    successful_hash = true;
+
+                                    // Now that the integrity hash was verified,
+                                    // attempt decryption.
+                                    //
+                                    // Note: decrypt_in_place does not overwrite
+                                    // the buffer if it fails, otherwise we
+                                    // would need to copy here. There's a unit
+                                    // test to validate this behaviour.
+                                    let decryption_result = context
+                                        .decrypt_in_place(
+                                            &mut response.data[..],
+                                            Nonce::from_slice(&ctx.nonce[..]),
+                                            Tag::from_slice(&ctx.tag[..]),
+                                        );
+
+                                    if decryption_result.is_ok() {
+                                        successful_decryption = true;
+                                        break;
+                                    }
                                 }
                             }
 
-                            if !successful_decryption {
-                                decryption_error =
-                                    Some(Err(CrucibleError::DecryptionError));
+                            if !successful_hash {
+                                // no hash was correct
+                                error = Some(CrucibleError::HashMismatch);
                                 break;
+                            } else if !successful_decryption {
+                                // no encryption context decrypted this block
+                                error = Some(CrucibleError::DecryptionError);
+                                break;
+                            } else {
+                                // Ok!
                             }
                         } else {
-                            // No encryption context! Either this is a read of
-                            // an unwritten block, or an attacker removed the
-                            // encryption contexts from the db
+                            // No encryption context in the response!
+                            //
+                            // Either this is a read of an unwritten block, or
+                            // an attacker removed the encryption contexts from
+                            // the db.
+                            //
+                            // XXX if it's not a blank block, we may be under
+                            // attack?
+                            assert!(response.data[..].iter().all(|&x| x == 0));
```

Comment:
buildomat's error is because of this line:

```
102165 | 2022-01-14T18:49:12.095Z | thread 'crucible-tokio' panicked at 'assertion failed: response.data[..].iter().all(\|&x\| x == 0)', upstairs/src/lib.rs:1613:29
-- | -- | --
102166 | 2022-01-14T18:49:12.102Z | note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I wasn't right with my reproduction using the "one" workload, but I believe this occurs because of a similar scenario: an unencrypted write, followed by an encrypted read. I think hammer triggers this because it's using the Crucible pseudo file and RMW is performing this encrypted read.

Reply:
This is no longer broken, right?  I can run `test_up.sh encrypted` on my clone of your repo and it passes.